# SQL-komentojen kutsuminen Pythonista

Pääkysymys:
    Miten hallita SQL-tietokantaa Python-ohjelmasta?

Mitä käsitellään?
    Python modulia nimelä `sqlite`.

Mitä sinun oletetaan tekevän?
    Lue ohjeet ja tee tehtävät.

Suuntaa antava vaativuusarvio:
    Helpohko.

Suuntaa antava työläysarvio:
    2-3 tuntia.

Ohjelmointitehtävät:
    Tehtävät ja niiden testit löytyvät omista hakemistoistaan.

## 1. Python kertaus

Aloitetaan lyhyellä Python-ohjelmoinnin kertauksella.
Tarvitset tehtävissä Pythonin perusrakenteiden lisäksi
mahdollisuuden avata, lukea ja sulkea tiedostoja.
Jos aihepiiri on entuudestaan tuttua, voit siirtyä suoraan
lukuun 2.

Pythonilla voidaan helposti lukea tietoa tiedostoista, sekä tallentaa niihin tietoa
useissa eri tiedostoformaateissa, kuten puhdas tekstitiedosto, `csv`, `xml`, `JSON` jne.
Tällä kurssilla hyödynnämme vain tekstitiedostoja (`.txt`).

Ennen tiedoston lukua se täytyy avata tiedosto `open()`-funktiolla, joka
palauttaa kahvan tiedostoon (esim. muuttuja `file`).

```Python
file = open(filename, mode)```

Funktion parametrina annetaan merkkijonona tiedoston nimi (`filename`),
sekä moodi (`mode`), joka kertoo avataanko tiedosto lukemista `'r'` (read) vai
kirjoittamista `'w'` (write) varten.

```Python
filename = 'file.txt'
file = open(filename, 'r')```

Moodin voi jättää pois, jolloin tiedosto avataan lukemista varten.

Tiedosto tulee lopuksi sulkea metodikutsulla `file.close()`.

### Tiedoston lukeminen rivi kerrallaan

Tiedostoa voi lukea rivi kerrallaan kutsumalla metodia `readline()`
```Python
row = file.readline()```

Koetetaan esimerkkinä lukea tehtävän 2 tiedostoa
`suomen_suurimpien_kuntien_asukasluvut.txt`

In [9]:
file = open('teht2/suomen_suurimpien_kuntien_asukasluvut.txt','r')
for row in file: # luetaan rivi kerrallaan
    print(row)
file.close()

Suurimpien kuntien väkiluku 31.12.2016. Lähde: Tilastokeskus [1.8.2017], http://www.tilastokeskus.fi/tup/suoluk/suoluk_vaesto.html

Helsinki;635181

Espoo;274583

Tampere;228274

Vantaa;219341

Oulu;200526

Turku;187604

Jyväskylä;138850

Lahti;119452

Kuopio;117740

Kouvola;85306

Pori;85059

Joensuu;75848

Lappeenranta;72872

Hämeenlinna;67850

Vaasa;67620

Rovaniemi;62231

Seinäjoki;62052

Mikkeli;54517

Kotka;54187

Salo;53546

Porvoo;50144

Kokkola;47723

Lohja;47149

Hyvinkää;46596

Nurmijärvi;42010

Järvenpää;41529

Rauma;39614

Kirkkonummi;39033

Tuusula;38588

Kajaani;37521

Kerava;35511

Savonlinna;35242



Tiedostosta luettu rivi on merkkijono,
jota on usein tarpeen jakaa pienempiin osiin esim. `split()`-metodilla.
Parameterinä sille annetaan erotin.
Tehtävässä 2 tulee tietueen eri kentät erottaa puolipisteellä.
Kokeile muokata em. koodia siten, että se erottaa tiedoston kaksi eri
kenttää omiin muuttujiin `nimi` ja `populaatio`. Tulosta sitten nämä
muuttujat. Laske myös populaatiot yhteen ja tulosta lopuksi
saatu luku (3363299). Huom! tiedoston ensimmäinen rivi, joka on kommentti,
tulee jättää huomioimatta. Kopio yllä oleva koodi tähän alle ja muokkaa sitä:

## 2. SQL ja Python

Kaikissa tehtävissä käytetään alla olevaa tietokantaa, johon on ladattu suuri määrä maailman kaupunkeja ja niiden asukasmääriä ja koordinaatteja. Tietokannassa on vain yksi taulu, jonka tietueet kuvaavat maailman kaupunkeja.

Ennen tehtävien aloittamista kannattaa tutkia tietokannan tietoja. Missä muodossa kaupunkien ja valtioiden nimet on annettu?

TABLE kaupungit

        id           integer primary key
        nimi         text
        alue         text
        valtio       text
        populaatio   integer
        lat          real
        lon          real

### SQL-komennot Python-ohjelmassa

Python-ohjelmista voidaan suorittaa SQLite-tietokannan komentoja lataamalla kirjasto `sqlite3`:

In [10]:
import sqlite3

Kanta avataan komennolla `connect`:


In [11]:
conn = sqlite3.connect("kaupungit.db")

Lisäksi tarvitaan ns. tietokantakursori, joka tässä yhteydessä antaa tietokannalle komentoja ja
vastaanottaa siltä tietoa:

In [12]:
c = conn.cursor()

Nyt voitaisiin suorttaa esimerkiksi seuraavanlainen SQL-komento `c.execute()`-metodilla:

In [13]:
c.execute("""SELECT * FROM kaupungit LIMIT 10""")

OperationalError: no such table: kaupungit

Saat komennon palauttaman ensimmäisen tietueen muuttujaan `item` seuraavasti

In [ ]:
item = c.fetchone()

Saat kaikki komennon palauttamat tietueet talteen komennolla

In [ ]:
list = c.fetchall()  #palauttaa kaikki c.excecute()-komennon palauttamat tietueet listana


Harjoitus: Kokeile tulostaa listasta `list` kaupungit ja niiden populaatiot.
Laske yhteen ja tulosta myös kymmenen ensimmäisen kaupungin populaatio. (2353589).

Komentoja on helppo formatoida, jolloin osa komennon parametreista voidaan määritellä muuttujissa.
Kysymysmerkit korvataan annetulla listalla muuttujia samassa järjestyksessä.

In [ ]:
nimi = "Kauniainen"
alue = "Southern Finland"
valtio = "FINLAND"
populaatio = 9960
lat = 60.21
lon = 24.66
c.execute("""INSERT INTO kaupungit(nimi, alue, valtio, populaatio, lat, lon)
                 VALUES(?,?,?,?,?,?)""",
                 [nimi, alue, valtio, populaatio, lat, lon])

Jos tietokantaan tehdään muutoksia (`UPDATE`, `INSERT`, `DELETE`), ne täytyy vielä kommitoida. Lisäksi
on hyvä tapa aina sulkea tietokanta lopuksi. Päivitykset on hyvä tehdä `try`-lohkossa, jonka
päätteeksi tietokanta suljetaan.

In [ ]:

    try:
        # Päivitetään tiedot
        c.execute("UPDATE kaupungit SET populaatio = ? WHERE nimi = ?", [635181, "Helsinki"])
        # Tallennetaan tehdyt muutokset tietokantaan
        conn.commit()

    finally:
        conn.close()

Kokeile avata tietokanta uudelleen ja tarkista tallentuiko Helsingin
päivitetty asukasluku?

### Vihjeitä

Jos seuraavissa tehtävissä konsoliin tulostuu virheilmoitus, jossa kerrotaan, että tietokantaa tai jotain taulua ei löydy,

```Python
    Traceback (most recent call last):

        File "kaupunkihaku.py", line 56, in <module>
            main()
        File "kaupunkihaku.py", line 38, in main
            queryresult = find_city_by_name(inputstring, tietokanta)

        File "kaupunkihaku.py", line 14, in find_city_by_name c.execute(sqlitecommand, [city_name])

    sqlite3.OperationalError: no such table: kaupungit```

kannattaa antaa `main()`-funktion `tietokanta`-muuttujaan tiedostopolku, jossa tietokanta sijaitsee.
Näin tietokantaa ei tarvitse kopioida tehtäväkohtaisiin alihakemistoihin.

```Python
    tietokanta = '../kaupungit.db'```

Huom! Tietokanta luodaan automaattisesti, jos sitä ei löydy. Tietokanta on tällöin tyhjä (0 tavua), joten vaikka näyttäisi siltä, että tietokanta on oikeassa paikassa, se ei toimi.

    ~/teht1$ ls -al
    total 24
    drwxr-xr-x 1 jovyan jovyan 4096 Dec  9 15:09 .
    drwxr-xr-x 1 jovyan jovyan 4096 Dec  9 15:08 ..
    -rw-r--r-- 1 jovyan root      0 Dec  9 14:34 kaupungit.db
    -rw-r--r-- 1 jovyan jovyan 1242 Dec  9 14:29 kaupunkihaku.py
    -rw-r--r-- 1 jovyan jovyan 4345 Dec  9 14:29 test_kaupunkihaku.py
    ~/teht1$

## Tehtävä 1.7: Kaupunkihaku

Täydennä Python-ohjelmaa `kaupunkihaku.py`, joka etsii tietokannasta nimen perusteella kaupungin ja tulostaa sen nimen, alueen, valtion, asukasmäärän sekä koordinaatit. Lue olemassa oleva koodi ja sen kommentointi ennen kuin alat kirjoittaa ratkaisua.

Huomaa, että kaupungin nimellä hakiessa voi tulla useita tuloksia.
Tällöin halutaan se kaupunki, jossa on eniten asukkaita.
Helpoin tapa valita väkimäärän perusteella on lisätä hakukomentoon
`ORDER BY populaatio DESC` eli järjestä laskevasti väkimäärän mukaan.
Tällöin suurin kaupunki on listassa ensimmäinen.

## Tehtävä 1.8: Tietojen päivittäminen

Ennen seuraavan tehtävän aloittamista kannattaa tutkia tehtäväpaketin mukana tulleita tiedostoja. Mitä merkistöä esimerkiksi tekstitiedosto kaupungit.txt käyttää? Ongelmien välttämiseksi tiedoston avaamisen yhteyteen on syytä lisätä tieto käytetystä merkistöstä. Python 3:ssa se onnistuu lisäämällä määre encoding esimerkiksi seuraavasti

```Python
    f = open(filename, 'r', encoding='utf-8')```

Tehtäväpaketin mukana tulee tiedosto `suomen_suurimpien_kuntien_asukasluvut.txt`,
jossa on listattuna Suomen suurimpien kuntien asukaslukuja. Tehtävänäsi on tiedostoa hyödyntämällä päivittää kaupunkien asukasluvut. Jos tiedostossa olevaa kaupunkia ei löydy tietokannasta, ohita kyseinen kaupunki. Älä kuitenkaan luo uutta kaupunkia tietokantaan.

## Tehtävä 1.9: Kahden kaupungin etäisyys

Tee Python-ohjelma, joka etsii kaksi kaupunkia tietokannasta nimen perusteella ja laskee niiden välisen etäisyyden.
Voit käyttää apunasi edellä tehtyä kaupunkihakua.
Huomioi, että samannimisistä kaupungeista halutaan se, jolla on suurin asukasluku.

Vihje. Etäisyys koordinaattien välillä kannattaa laskea isoympyrän kulman avulla.
Pisteiden `(lon1,lat1)` ja `(lon2,lat2)` väliselle etäisyydelle saadaan seuraavat yhtälöt,
jos Maapallon säde on `R`. Muista pohtia, oletko käyttämässä radiaaneja vai asteita.
Käytä ratkaisussasi math-kirjaston funktioita, testiohjelma ei lataa esim. numpyä.
kaavat etäisyyden laskemiseen isoympyrän avulla

   Kulma isoympyrällä:

   $\alpha = \arccos(\sin(lat1) * \sin(lat2) + \cos(lat1) * cos(lat2) * cos(lon2-lon1))$

   Etäisyys isoympyrällä:

   $b = \alpha * R$

Miten voin testata tehtävien toimivuutta?

- Etsi SQLite DB Browserilla joku kaupunki tietokannasta ja tutki löytääkö ohjelma sen.
- Tietojen päivittämisen jälkeen tutki SQLite DB Browserilla ovatko tiedot päivittyneet tietokantaan.
- Laske esimerkiksi laskimella oikea tulos ja vertaa sitä ohjelmasi palauttamaan tulokseen.